src/linear_algebra/matrix.lean

```diff
@@ -972,3 +974,121 @@ lemma matrix.dot_product_eq_zero_iff {v : n → R} : (∀ w, matrix.dot_product
 ⟨λ h, matrix.dot_product_eq_zero v h, λ h w, h.symm ▸ zero_dot_product w⟩
 
 end
+
+namespace matrix
+
+variables {m n : Type u} [decidable_eq n] [fintype n] [decidable_eq m] [fintype m] {R : Type v}
+  [comm_ring R]
+
+lemma to_block_matrix_det (M : matrix m m R) (p : m → Prop) [decidable_pred p] :
+  M.det = (matrix.from_blocks (to_block M p p) (to_block M p (λ j, ¬p j))
+    (to_block M (λ j, ¬p j) p) (to_block M (λ j, ¬p j) (λ j, ¬p j))).det :=
+begin
+  rw ← matrix.det_reindex_self (equiv.sum_compl p).symm M,
+  unfold det,
+  congr, ext σ, congr, ext,
+  generalize hy : σ x = y,
+  cases x; cases y;
+  simp only [matrix.reindex_apply, to_block_apply, equiv.symm_symm,
+    equiv.sum_compl_apply_inr, equiv.sum_compl_apply_inl,
+    from_blocks_apply₁₁, from_blocks_apply₁₂, from_blocks_apply₂₁, from_blocks_apply₂₂],
+end
+
+lemma to_square_block_det (M : matrix m m R) (b : m → ℕ) (k : ℕ) :
+  (to_square_block M b k).det = (to_square_block' M (λ i, b i = k)).det := by simp
+
+lemma upper_two_block_triangular_det' (M : matrix m m R) (p : m → Prop) [decidable_pred p]
+  (h : ∀ i (h1 : ¬p i) j (h2 : p j), M i j = 0) :
+  M.det = (to_square_block' M p).det * (to_square_block' M (λ i, ¬p i)).det :=
+begin
+  rw to_block_matrix_det M p,
+  convert upper_two_block_triangular_det (to_block M p p) (to_block M p (λ j, ¬p j))
+    (to_block M (λ j, ¬p j) (λ j, ¬p j)),
+  ext,
+  exact h ↑i i.2 ↑j j.2
+end
+
+lemma equiv_block_det (M : matrix m m R) {p q : m → Prop} [decidable_pred p] [decidable_pred q]
+  (e : ∀x, q x ↔ p x) : (to_square_block' M p).det = (to_square_block' M q).det :=
+by convert matrix.det_reindex_self (equiv.subtype_equiv_right e) (to_square_block' M q)
+
+/-- Let `b` map rows and columns of a square matrix `M` to `n` blocks. Then
+  `upper_block_triangular_matrix M n b` says the matrix is upper block triangular. -/
+def upper_block_triangular_matrix {o : Type*} [fintype o] (M : matrix o o R) (n : ℕ)
```

Comment:
Should this be changed so that `b : o → p` where `p` is an _ordered_ fintype? I think the current version might be easier to use, but I think the ordered fintype version would mean we don't need to eventually add a lower_block_triangular...

Reply:
Requiring `linear_order p` sounds like a good idea to me, since as you remark you can then use `order_dual p` to get the lower version.